### PR TITLE
Added MangaHere source

### PR DESCRIPTION
**Pull request by MareGraphics**
I added all basic functionality for MangaHere. 
There's still some work to be done, but it's all minor stuff.
Screenshots:

### CODE REVIEW

File: websites/M/MangaHere/presence.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    if (pathname.endsWith("/on_going/")) {
+      var url = pathname;
+      var splitUrl = url.split('/');
+      if (splitUrl[1] == "on_going") {
+        data.details = "Browsing ongoing manga";
+      } else {
+        data.details = "Browsing ongoing " + splitUrl[1] + " manga";
+      }
+      data.startTimestamp = browsingStamp;
+    }
```
Comment (Bas950) [inline websites/M/MangaHere/presence.ts:83]: else if?

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/M/MangaHere/presence.ts (hunk 1 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:83]: ... I accidentally deleted it all

File: websites/M/MangaHere/dist/metadata.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    "category": "anime",
+    "logo": "https://i.imgur.com/w7zc3W4.png",
+    "thumbnail": "https://i.imgur.com/yUGRcWL.png",
+    "version": "1.0.0",
+    "color": "#3d6eff",
+    "tags": [
+      "manga",
+      "manhwa"
+    ],
+    "url": ["www.mangahere.cc","mangahere.cc"],
```
Comment (EncryptedDev) [inline websites/M/MangaHere/dist/metadata.json:17]: You don't need both URLs?

File: websites/M/MangaHere/dist/metadata.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+    "$schema": "https://schemas.premid.app/metadata/1.3",
+    "service": "MangaHere",
+    "author": {
+      "name": "Harold",
+      "id": "141206144264699904"
+    },
+    "category": "anime",
+    "logo": "https://i.imgur.com/w7zc3W4.png",
```
Comment (EncryptedDev) [inline websites/M/MangaHere/dist/metadata.json:9]: Not really necessary, but the quality of that logo is just...

File: websites/M/MangaHere/presence.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+              "sci-fi", "yoi", "shotacon", "mystery", "shoujo",
+              "ecchi", "doujinshi", "lolicon", "adventure", "romance",
+              "gender-bender", "harem", "sports", "webtoons", "comedy",
+              "shounen-ai", "josei", "shoujo-ai", "adult", "fantasy", "supernatural",
+              "psychological", "yuri", "one-shot", "historical", "drama",
+              "seinen", "mature", "smut", "horror", "shounen", "slice-of-life",
+              "tragedy", "mecha"
+            ];
+
+  if (hostname === "mangahere.cc" || hostname === "www.mangahere.cc") {
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:24]: That's the same thing in different variations lol, just include one

File: websites/M/MangaHere/presence.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      if (splitUrl[1] == "new") {
+        data.details = "Browsing new manga";
+      } else {
+        data.details = "Browsing new " + splitUrl[1] + " manga";
+      }
+      data.startTimestamp = browsingStamp;
+    }
+    //ganre/completed/
+    else if (pathname.endsWith("/completed/")) {
+      let splitUrl = url.split('/');
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:63]: Your consistency of using const or let is bad, why don't you just stick to const as the value is never reassigned?

File: websites/M/MangaHere/presence.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        data.details = "Browsing ongoing " + splitUrl[1] + " manga";
+      }
+      data.startTimestamp = browsingStamp;
+    }
+    
+
+    //Manga Viewing
+    else if (pathname.startsWith("/manga") && pathname.endsWith("/")) {
+      const title = document.querySelector(".detail-info-right-title-font").textContent;
+      let link = window.location.href;
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:87]: const as the value is never reassigned?

File: websites/M/MangaHere/presence.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      let chapter = document.querySelector(".reader-header-title-2").textContent;
+      //setting up page progress
+      let current = document.querySelector('.pager-list-left span');
+      if(current == null){
+        data.state = chapter;
+      }else{
+        let len = current.children.length;
+        let totalPages = current.children[len - 2].textContent;
+        let readingPage = document.querySelector(".pager-list-left > span > .active").textContent;
+        let progress = readingPage + "/" + totalPages;
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:107]: merge lets?

File: websites/M/MangaHere/presence.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      data.state = title;
+      data.startTimestamp = browsingStamp;
+      data.buttons = [{ label: "Read", url: link }];
+      data.smallImageKey = "viewing";
+      data.smallImageText = "Viewing";
+    }
+    //Manga Reading
+    else if (pathname.startsWith("/manga") && pathname.endsWith(".html")) {
+      let title = document.querySelector(".reader-header-title-1").textContent;
+      let chapter = document.querySelector(".reader-header-title-2").textContent;
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:98]: merge lets?

File: websites/M/MangaHere/presence.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        data.details = "Browsing completed manga";
+      } else {
+        data.details = "Browsing completed " + splitUrl[1] + " manga";
+      }
+      data.startTimestamp = browsingStamp;
+    }
+    //ganre/on_going/
+    else if (pathname.endsWith("/on_going/")) {
+      var url = pathname;
+      var splitUrl = url.split('/');
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:74]: why use var, const, let all in one file? Just be more consistent?

File: websites/M/MangaHere/presence.ts [hunk 10]
```diff
@@ -0,0 +1,144 @@
+const presence = new Presence({
+  clientId: "831262912815300638"
+}),
+  browsingStamp = Math.floor(Date.now() / 1000);
+
+// let username: string;
```
Comment (EncryptedDev) [inline websites/M/MangaHere/presence.ts:6]: Remove the comment as it's not used?

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
Hello, look at my review. 
- Please fix your consistency of variable usage

File: websites/M/MangaHere/dist/metadata.json (hunk 2 above)
Comment (MareGraphics) [reply websites/M/MangaHere/dist/metadata.json:17]: I did it just in case, I'll remove it.

File: websites/M/MangaHere/dist/metadata.json (hunk 3 above)
Comment (MareGraphics) [reply websites/M/MangaHere/dist/metadata.json:9]: Ugh, I wanted to recreate it, but it's too much work, they just don't have HD logo..

File: websites/M/MangaHere/presence.ts (hunk 4 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:24]: I did it just in case xD I'll fix it now.

File: websites/M/MangaHere/presence.ts (hunk 6 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:87]: Will fix it.

File: websites/M/MangaHere/presence.ts (hunk 7 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:107]: Sure, just got too used to php and stuff so getting used to ts again

File: websites/M/MangaHere/presence.ts (hunk 8 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:98]: Roger

File: websites/M/MangaHere/presence.ts (hunk 9 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:74]: Yeah, that's what I'm fixing right now, my variables are all over the place

File: websites/M/MangaHere/presence.ts (hunk 10 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:6]: Roger!

File: websites/M/MangaHere/presence.ts (hunk 5 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:63]: Yeah i just became aware of it, I will stick to const.

**Review by EncryptedDev — COMMENTED** [review]
Cool, looks fine to me. You should use prettier to format your code though

File: websites/M/MangaHere/dist/metadata.json (hunk 3 above)
Comment (ririxidev) [reply websites/M/MangaHere/dist/metadata.json:9]: Just resize their logo to 512x512 ty.

File: websites/M/MangaHere/presence.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      data.startTimestamp = browsingStamp;
+    }
+    else if (pathname == "/directory/") {
+      data.details = "Browsing all manga";
+      data.startTimestamp = browsingStamp;
+    }
+    else if (pathname == "/on_going/") {
+      data.details = "Browsing ongoing manga";
+      data.startTimestamp = browsingStamp;
+    }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:48]: `==` -> `===`

File: websites/M/MangaHere/presence.ts [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+              "seinen", "mature", "smut", "horror", "shounen", "slice-of-life",
+              "tragedy", "mecha"
+            ];
+
+  if (hostname === "www.mangahere.cc") {
+
+    if (pathname == "/") {
+      data.details = "Viewing the Homepage";
+      data.startTimestamp = browsingStamp;
+      data.buttons = [{ label: "Visit the website", url: "www.mangahere.cc" }];
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:27]: image

File: websites/M/MangaHere/presence.ts [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      data.startTimestamp = browsingStamp;
+    }
+    //Manga Viewing
+    else if (pathname.startsWith("/manga") && pathname.endsWith("/")) {
+      const title = document.querySelector(".detail-info-right-title-font").textContent,
+       link = window.location.href;
+      data.details = "Viewing manga:";
+      data.state = title;
+      data.startTimestamp = browsingStamp;
+      data.buttons = [{ label: "Read", url: link }];
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:89]: View Manga

File: websites/M/MangaHere/presence.ts [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    //Manga Reading
+    else if (pathname.startsWith("/manga") && pathname.endsWith(".html")) {
+      const title = document.querySelector(".reader-header-title-1").textContent,
+       chapter = document.querySelector(".reader-header-title-2").textContent,
+      //setting up page progress
+       current = document.querySelector('.pager-list-left span');
+      if(current == null){
+        data.state = chapter;
+      }else{
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:101]:
```suggestion
      if(!current) data.state = chapter;
      else {
```

File: websites/M/MangaHere/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{
+    "extends": "../../../tsconfig.json",
+    "compilerOptions": {
+      "outDir": "./dist/"
+    }
+  }
+
```
Comment (ririxidev) [inline websites/M/MangaHere/tsconfig.json:7]: delete blank space.

File: websites/M/MangaHere/presence.ts [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+      data.smallImageText = "Searching";
+    }
+    //Browisng ganre
+    for(let i = 0; i<ganres.length; i++){
+      if(pathname.substring(1, pathname.length-1) == ganres[i]){
+        data.details = "Browsing:";
+        data.state = ganres[i].replace('-',' ') +" manga";
+      data.startTimestamp = browsingStamp;
+      }
+    }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:137]: maybe you could use ``.map`` or ``.forEach`` instead?

File: websites/M/MangaHere/presence.ts [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    else if (pathname.startsWith("/search")) {
+      const queryString = window.location.search,
+       urlParams = new URLSearchParams(queryString),
+       search = urlParams.get('title');
+       let searchName = "";
+      if (search == "") {
+        searchName = urlParams.get('name');
+      } else {
+        searchName = urlParams.get('title');
+      }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:123]:
```suggestion
       search = urlParams.get('title'),
       searchName = search === "" ? urlParams.get("name") : urlParams.get("title");
```

File: websites/M/MangaHere/presence.ts [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    //ganre/new/
+    else if (pathname.endsWith("/new/")) {
+      const url = pathname,
+       splitUrl = url.split('/');
+      if (splitUrl[1] == "new") {
+        data.details = "Browsing new manga";
+      } else {
+        data.details = "Browsing new " + splitUrl[1] + " manga";
+      }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:57]:
```suggestion
      data.details = splitUrl[1] === "new" ? "Browsing new manga" : `Browsing new ${splitUrl[1]} manga`
```

**Review by ririxidev — CHANGES_REQUESTED** [review]
Heya @MareGraphics!
You should check out thing(s) that I requested to change in my review. Please look at them!
 - you can move your all data.startTimestamp things to ``data`` object instead of setting it everytime :/

Regards,
**ririxi**. PreMiD Reviewer Team.
Contact: mail@ririxi.dev, ririxi#2721.

File: websites/M/MangaHere/dist/metadata.json (hunk 3 above)
Comment (Bas950) [reply websites/M/MangaHere/dist/metadata.json:9]: Yea try getting a better logo

File: websites/M/MangaHere/presence.ts [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+              "seinen", "mature", "smut", "horror", "shounen", "slice-of-life",
+              "tragedy", "mecha"
+            ];
+
+  if (hostname === "www.mangahere.cc") {
+
+    if (pathname == "/") {
+      data.details = "Viewing the Homepage";
+      data.startTimestamp = browsingStamp;
+      data.buttons = [{ label: "Visit the website", url: "www.mangahere.cc" }];
```
Comment (Bas950) [inline websites/M/MangaHere/presence.ts:27]:
```suggestion
```

File: websites/M/MangaHere/presence.ts [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+      data.startTimestamp = browsingStamp;
+    }
+    //Manga Viewing
+    else if (pathname.startsWith("/manga") && pathname.endsWith("/")) {
+      const title = document.querySelector(".detail-info-right-title-font").textContent,
+       link = window.location.href;
+      data.details = "Viewing manga:";
+      data.state = title;
+      data.startTimestamp = browsingStamp;
+      data.buttons = [{ label: "Read", url: link }];
```
Comment (Bas950) [inline websites/M/MangaHere/presence.ts:85]:
```suggestion
      data.buttons = [{ label: "View Manga", url: link }];
```

Comment (Bas950) [reply websites/M/MangaHere/presence.ts:85]:
```suggestion
      data.buttons = [{ label: "Read Manga", url: link }];
```

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/M/MangaHere/tsconfig.json (hunk 15 above)
Comment (MareGraphics) [reply websites/M/MangaHere/tsconfig.json:7]: Sure.

File: websites/M/MangaHere/presence.ts (hunk 13 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts]: What exactly do You mean?

File: websites/M/MangaHere/presence.ts (hunk 16 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:126]: Sure, I went with .forEach. Thank You

File: websites/M/MangaHere/presence.ts (hunk 13 above)
Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts]: Isn't it better to use "Viewing"?

File: websites/M/MangaHere/presence.ts [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    else if (pathname === "/spoilers/") {
+      data.details = "Browsing spoilers and news";
+    }
+    else if (pathname === "/directory/") {
+      data.details = "Browsing all manga";
+    }
+    else if (pathname === "/on_going/") {
+      data.details = "Browsing ongoing manga";
+    }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:42]: don't need brackets here

File: websites/M/MangaHere/presence.ts [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+              "ecchi", "doujinshi", "lolicon", "adventure", "romance",
+              "gender-bender", "harem", "sports", "webtoons", "comedy",
+              "shounen-ai", "josei", "shoujo-ai", "adult", "fantasy", "supernatural",
+              "psychological", "yuri", "one-shot", "historical", "drama",
+              "seinen", "mature", "smut", "horror", "shounen", "slice-of-life",
+              "tragedy", "mecha"
+            ];
+
+  if (hostname === "www.mangahere.cc") {
+
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:24]: remove blank line

File: websites/M/MangaHere/presence.ts [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    //ganre/completed/
+    else if (pathname.endsWith("/completed/")) {
+      const url = pathname,
+       splitUrl = url.split('/');
+      if (splitUrl[1] == "completed") {
+        data.details = "Browsing completed manga";
+      } else {
+        data.details = "Browsing completed " + splitUrl[1] + " manga";
+      }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:57]:
```suggestion
      data.details = splitUrl[1] === "completed" ? "Browsing completed manga" : `Browsing completed ${splitUrl[1]} manga`;
```

File: websites/M/MangaHere/presence.ts [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    //ganre/on_going/
+    else if (pathname.endsWith("/on_going/")) {
+      const url = pathname,
+       splitUrl = url.split('/');
+      if (splitUrl[1] == "on_going") {
+        data.details = "Browsing ongoing manga";
+      } else {
+        data.details = "Browsing ongoing " + splitUrl[1] + " manga";
+      }
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:67]:
```suggestion
      data.details = splitUrl[1] === "on_going" ? "Browsing ongoing manga" : `Browsing ongoing ${splitUrl[1]} manga`;
```

File: websites/M/MangaHere/presence.ts [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    //Manga Viewing
+    else if (pathname.startsWith("/manga") && pathname.endsWith("/")) {
+      const title = document.querySelector(".detail-info-right-title-font").textContent,
+       link = window.location.href;
+      data.details = "Viewing manga:";
+      data.state = title;
+      data.buttons = [{ label: "View Manga", url: link }];
+      data.smallImageKey = "viewing";
+      data.smallImageText = "Viewing";
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:77]:
```suggestion
      data.smallImageKey = "viewing";
```

File: websites/M/MangaHere/presence.ts [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+      else {
+         const len = current.children.length,
+         totalPages = current.children[len - 2].textContent,
+         readingPage = document.querySelector(".pager-list-left > span > .active").textContent,
+         progress = readingPage + "/" + totalPages;
+        data.state = chapter + " page " + progress;
+      }
+      data.details = title;
+      data.smallImageKey = "reading";
+      data.smallImageText = "Reading";
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:95]:
```suggestion
      data.smallImageKey = "reading";
```

File: websites/M/MangaHere/presence.ts [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+    //Searching
+    else if (pathname.startsWith("/search")) {
+      const queryString = window.location.search,
+       urlParams = new URLSearchParams(queryString),
+       search = urlParams.get('title'),
+       searchName = search === "" ? urlParams.get("name") : urlParams.get("title");
+      data.details = "Searching:";
+      data.state = searchName;
+      data.smallImageKey = "searching";
+      data.smallImageText = "Searching";
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:106]:
```suggestion
      data.smallImageKey = "searching";
```

**Review by ririxidev — CHANGES_REQUESTED** [review]

File: websites/M/MangaHere/presence.ts [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+              "sci-fi", "yoi", "shotacon", "mystery", "shoujo",
+              "ecchi", "doujinshi", "lolicon", "adventure", "romance",
+              "gender-bender", "harem", "sports", "webtoons", "comedy",
+              "shounen-ai", "josei", "shoujo-ai", "adult", "fantasy", "supernatural",
+              "psychological", "yuri", "one-shot", "historical", "drama",
+              "seinen", "mature", "smut", "horror", "shounen", "slice-of-life",
+              "tragedy", "mecha"
+            ];
+
+  if (hostname === "www.mangahere.cc") {
```
Comment (ririxidev) [inline websites/M/MangaHere/presence.ts:23]: I just noticed that you don't need this `if` statement since you have only one domain specified in metadata.

Comment (MareGraphics) [reply websites/M/MangaHere/presence.ts:23]: Oh, exactly, i didn't delete it since I had 2 domains at first. On it :)

**Review by ririxidev — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**